pkg/uri/uri.go

```diff
@@ -3,75 +3,104 @@ package uri
 import (
 	"io"
 	"net/http"
+	"net/url"
 	"os"
 	"regexp"
-	"strings"
 
 	"github.com/square/p2/pkg/util"
 )
 
-var leadingProto = regexp.MustCompile("^[a-zA-Z\\d\\.]+:.*")
+// Internet Standard: STD 66
+var leadingScheme = regexp.MustCompile(`^([a-zA-Z][a-zA-Z0-9+\-.]*):`)
 
-type Copier struct {
-	Client *http.Client
-}
-
-func NewCopier(client *http.Client) *Copier {
-	copier := &Copier{
-		Client: client,
-	}
+// A UriFetcher presents simple methods for fetching URIs from
+// different schemes.
+type Fetcher interface {
+	// Opens a data stream to the source URI. If no URI scheme is
+	// specified, treats the URI as a path to a local file.
+	Open(uri string) (io.ReadCloser, error)
 
-	return copier
+	// Copy all data from the source URI to a local file at the
+	// destination path.
+	CopyLocal(srcUri, dstPath string) error
 }
 
-var DefaultCopier *Copier = NewCopier(http.DefaultClient)
+// A default fetcher, if the user doesn't want to set any options.
+var DefaultFetcher Fetcher = BasicFetcher{http.DefaultClient}
 
 // URICopy Wraps opening and copying content from URIs. Will attempt
 // directly perform file copies if the uri is begins with file://, otherwise
 // delegates to a curl implementation.
-func URICopy(fromURI, toPath string) error {
-	return DefaultCopier.Copy(fromURI, toPath)
+var URICopy = DefaultFetcher.CopyLocal
+
+// BasicFetcher can access "file" and "http" schemes using the OS and
+// a provided HTTP client, respectively.
+type BasicFetcher struct {
+	Client *http.Client
+}
+
+func (f BasicFetcher) Open(srcUri string) (io.ReadCloser, error) {
+	uri, err := url.Parse(srcUri)
+	if err != nil {
+		return nil, util.Errorf("%s: bad URI", srcUri)
+	}
+	switch uri.Scheme {
+	case "file", "":
```

Comment:
Does `url.Parse` correctly return the empty string for the scheme if it's just a file path?